README.md
parse-superagent
================

A Parse.com REST client using  superagent.

[![Build Status](https://travis-ci.org/roothybrid7/parse-superagent.png)](https://travis-ci.org/roothybrid7/parse-superagent) [![NPM version](https://badge.fury.io/js/parse-superagent.png)](http://badge.fury.io/js/parse-superagent) [![Dependency Status](https://gemnasium.com/roothybrid7/parse-superagent.png)](https://gemnasium.com/roothybrid7/parse-superagent)


## Install

```
npm install parse-superagent
```

Additional Install for Parse cloud code end-to-end testing.

```
npm install supertest
```

## Usage

Simple example, included as `examples/client.js`:

```javascript
/**
 * client.js
 */

var parseSuperAgent = require('../lib'),
    events = require('events'),
    eventEmitter = new events.EventEmitter();

// Setup the parse keys by environment variables.
// export NODE_PARSE_APP_ID=<Application ID>
// export NODE_PARSE_REST_API_KEY=<REST API Key>
// export NODE_PARSE_MASTER_KEY=<Master Key>
// export NODE_USE_PARSE_MASTER_KEY=(0|1)

// Or the instantiate arguments.
var client = new parseSuperAgent(/*{
  applicationId: '<Application ID>',
  restApiKey: '<REST API Key>',
  masterKey: '<Master Key>',
  useMasterKey: true  # Or false
}*/);

client.runFunction('hello').end(function(error, response) {
  if (error) {
    throw error;
  }
  console.log(response.body);
  // => { result: 'Hello world!' }
});

client.sendBatchRequest({
          requests: [{
            method: 'POST',
            path: '/1/classes/GameScore',
            body: {
              score: 1337,
              playerName: "Sean Plott"
            }
          }, {
            method: 'POST',
            path: '/1/classes/GameScore',
            body: {
              score: 1338,
              playerName: "ZeroCool"
            }
          }]
}).end(function(error, response) {
  if (error) {
    throw error;
  }
  console.log(response.body);
  // => [ { success:
  // { createdAt: '2013-11-17T14:43:40.296Z',
  // objectId: 'aGFMHhdHBY' } },
  // { success:
  // { createdAt: '2013-11-17T14:43:40.308Z',
  // objectId: 'rpQbB6RXgT' } } ]
  eventEmitter.emit('batchEnd');
});

client.send('POST', '/1/classes/GameScore', {
  score:1337,
  playerName: "Sean Plott",
  cheatMode:false
}).end(function(error, response) {
  if (error) {
    throw error;
  }
  console.log(response.body);
  // => { createdAt: '2013-11-17T14:43:40.297Z',
  // objectId: 'W2tdhLOWKC' }
  eventEmitter.emit('created', response.body.objectId);
});

eventEmitter.on('created', function(objectId) {
  client
    .send('GET', '/1/classes/GameScore/' + objectId)
    .end(function(error, response) {
      if (error) {
        throw error;
      }
      console.log(response.body);
      // => { cheatMode: false,
      // playerName: 'Sean Plott',
      // score: 73453,
      // createdAt: '2013-11-17T14:43:40.297Z',
      // updatedAt: '2013-11-17T14:43:41.432Z',
      // objectId: 'W2tdhLOWKC' }
      eventEmitter.emit('fetched', response.body.objectId);
    });

  client
    .send('PUT', '/1/classes/GameScore/' + objectId, {score: 73453})
    .end(function(error, response) {
      if (error) {
        throw error;
      }
      console.log(response.body);
      // => { updatedAt: '2013-11-17T14:43:41.432Z' }
      eventEmitter.emit('updated', objectId);
    });
});

eventEmitter.on('updated', function(objectId) {
  client
    .send('DELETE', '/1/classes/GameScore/' + objectId)
    .end(function(error, response) {
      if (error) {
        throw error;
      }
      console.log(response.body);
      // => {}
      eventEmitter.emit('deleted', response.body);
    });
});

eventEmitter.on('deleted', function() {
  client
    .send('GET', '/1/classes/GameScore')
    .end(function(error, response) {
      if (error) {
        throw error;
      }
      console.log(response.body);
      // => { results:
      // [ { score: 1337,
      // playerName: 'Sean Plott',
      // createdAt: '2013-11-17T14:43:40.296Z',
      // updatedAt: '2013-11-17T14:43:40.296Z',
      // objectId: 'aGFMHhdHBY' },
      // { score: 1338,
      // playerName: 'ZeroCool',
      // createdAt: '2013-11-17T14:43:40.308Z',
      // updatedAt: '2013-11-17T14:43:40.308Z',
      // objectId: 'rpQbB6RXgT' } ] }
      //     });
      var requests = [];
      var results = response.body.results;
      for (var i = 0, l = results.length; i < l; i++) {
        requests.push({
          method: 'DELETE',
          path: '/1/classes/GameScore/' + results[i].objectId
        });
      }
      eventEmitter.emit('fetchAll', requests);
    });
});

eventEmitter.on('fetchAll', function(requests) {
  // Cleanup.
  client
    .sendBatchRequest({requests: requests})
    .end(function(error, response) {
      if (error) {
        throw error;
      }
      console.log(response.body);
      // => [ { success: true }, { success: true } ]
    });
});
```

Setup the Parse cloud code project by parse command line tool.

```
$ parse new
[...]
Email: foo@example.com
Password:
1: sampleApp
Select an App: 1
```

This will display:

```
$ node examples/client.js

# with NODE_ENV=testing(Use the `supertest` that is installed.)
$ NODE_ENV=testing examples/client.js
```


History.md

0.0.3 / 2013-11-18 
==================

 * Added NODE_ENV for Travis CI build

0.0.1 / 2013-11-18 
==================

 * initial release


test/helper.js
/**
 * helper.js - Test helper.
 * node v0.10.x
 */
(function(root, undefined) {
  'use strict';

  var path = require('path'),
      chai = require('chai'),
      sinon = require('sinon');

  chai.use(require('sinon-chai'));

  // Set global variables.
  global.expect = chai.expect;
  global.sinon = sinon;

  // Helper plugins.
  module.exports = {
  };
}(this));


test/lib/index_spec.js
/**
 * index_spec.js
 */
(function(root, undefined) {
  'use strict';

  var helper = require('../helper'),
      TargetModule = require('../../lib');

  describe('ParseSuperAgent', function() {
    var keys;

    before(function() {
      keys = {
        applicationId: 'APP_ID',
        restApiKey: 'REST_API_KEY',
        masterKey: 'MASTER_KEY'
      };
    });

    describe('#initialize', function() {
      context('input the parse keys', function() {
        var target;

        beforeEach(function() {
          var options = {};
          for (var k in keys) {
            options[k] = keys[k];
          }
          options.useMasterKey = true;
          target = new TargetModule(options);
        });

        afterEach(function() {
          target = null;
        });

        it('should be set the parse application keys', function() {
          expect(target).to.have.property('applicationId', keys.applicationId);
          expect(target).to.have.property('restApiKey', keys.restApiKey);
          expect(target).to.have.property('masterKey', keys.masterKey);
          expect(target).to.have.property('useMasterKey', true);
        });
      });

      context('setting env variables', function() {
        var target;

        beforeEach(function() {
          process.env.NODE_PARSE_APP_ID = 'APP_ID';
          process.env.NODE_PARSE_REST_API_KEY = 'REST_API_KEY';
          process.env.NODE_PARSE_MASTER_KEY = 'MASTER_KEY';
          process.env.NODE_USE_PARSE_MASTER_KEY = '1';
        });

        afterEach(function() {
          target = null;
          delete process.env.NODE_PARSE_APP_ID;
          delete process.env.NODE_PARSE_REST_API_KEY;
          delete process.env.NODE_PARSE_MASTER_KEY;
          delete process.env.NODE_USE_PARSE_MASTER_KEY;
        });

        context('no input parameters', function() {
          beforeEach(function() {
            target = new TargetModule();
          });

          it('should be set the parse application keys', function() {
            expect(target).to.have.property('applicationId', process.env.NODE_PARSE_APP_ID);
            expect(target).to.have.property('restApiKey', process.env.NODE_PARSE_REST_API_KEY);
            expect(target).to.have.property('masterKey', process.env.NODE_PARSE_MASTER_KEY);
            expect(target).to.have.property('useMasterKey', true);
          });
        });

        context('input `useMasterKey` with true', function() {
          beforeEach(function() {
            target = new TargetModule({useMasterKey: false});
          });

          it('set the parse application keys, but `useMasterKey` is false', function() {
            expect(target).to.have.property('applicationId', process.env.NODE_PARSE_APP_ID);
            expect(target).to.have.property('restApiKey', process.env.NODE_PARSE_REST_API_KEY);
            expect(target).to.have.property('masterKey', process.env.NODE_PARSE_MASTER_KEY);
            expect(target).to.have.property('useMasterKey', false);
          });
        });
      });
    });

    describe('#keySet', function() {
      var target;

      beforeEach(function() {
        target = new TargetModule(keys);
      });

      afterEach(function() {
        target = null;
      });

      context('undefined `useMasterKey`', function() {
        it('return the keyset without `X-Parse-Master-Key`', function() {
          var result = target.keySet();
          expect(result).to.have.property('X-Parse-Application-Id', keys.applicationId);
          expect(result).to.have.property('X-Parse-REST-API-Key', keys.restApiKey);
          expect(result).to.not.have.property('X-Parse-Master-Key');
        });
      });

      context('set `options.useMasterKey` to true', function() {
        it('return the keyset with `X-Parse-Master-Key`', function() {
          var result = target.keySet({useMasterKey: true});
          expect(result).to.have.property('X-Parse-Application-Id', keys.applicationId);
          expect(result).to.have.property('X-Parse-REST-API-Key', keys.restApiKey);
          expect(result).to.have.property('X-Parse-Master-Key', keys.masterKey);
          expect(target).to.have.deep.property('useMasterKey')
            .that.to.not.be.true;
        });
      });

      context('set the property of `useMasterKey` to true', function() {
        beforeEach(function() {
          target.useMasterKey = true;
        });

        it('return the keyset with `X-Parse-Master-Key`', function() {
          var result = target.keySet();
          expect(result).to.have.property('X-Parse-Application-Id', keys.applicationId);
          expect(result).to.have.property('X-Parse-REST-API-Key', keys.restApiKey);
          expect(result).to.have.property('X-Parse-Master-Key', keys.masterKey);
          expect(target).to.have.deep.property('useMasterKey')
            .that.to.be.true;
        });
      });
    });

    describe('#pathJoin', function() {
      var target, apiVer;

      beforeEach(function() {
        target = new TargetModule();
        apiVer = target.apiVer || '1';
      });

      it('set the parse rest api paths', function() {
        var result = target.pathJoin('/', apiVer, 'classes', 'GameScore');
        expect(result).to.be.eql('/' + apiVer + '/classes/GameScore');
      });
    });
  });
}(this));


Gruntfile.js
/*global module:false*/
module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    // Metadata.
    pkg: grunt.file.readJSON('package.json'),
    banner: '/*! <%= pkg.title || pkg.name %> - v<%= pkg.version %> - ' +
      '<%= grunt.template.today("yyyy-mm-dd") %>\n' +
      '<%= pkg.homepage ? "* " + pkg.homepage + "\\n" : "" %>' +
      '* Copyright (c) <%= grunt.template.today("yyyy") %> <%= pkg.author.name %>;' +
      ' Licensed <%= _.pluck(pkg.licenses, "type").join(", ") %> */\n',
    // Task configuration.
    jshint: {
      options: {
        jshintrc: '.jshintrc'
      },
      gruntfile: {
        src: 'Gruntfile.js'
      },
      js: {
        src: ['lib/**/*.js', 'test/**/*.js']
      }
    },
    mochaTest: {
      options: {
        reporter: 'mocha-unfunk-reporter',
        glowl: true
      },
      cloudUnit: {
        options: {
          bail: true
        },
        src: ['test/**/*_spec.js']
      },
    },
    watch: {
      files: ['<%= jshint.js.src %>'],
      tasks: ['default']
    }
  });

  // These plugins provide necessary tasks.
  var pkg = grunt.config.get('pkg');
  for (var name in pkg.devDependencies) {
    if (name.substring(0, 6) === 'grunt-') {
      grunt.loadNpmTasks(name);
    }
  }

  // Default task.
  grunt.registerTask('default', ['jshint', 'mochaTest']);

};


lib/index.js
/**
 * parse-superagent.js - Parse.com rest api client.
 */
(function(root, undefined) {
  'use strict';

  var superTest;
  try {
    superTest = require.resolve('supertest');
  } catch (ex) {
    superTest = null;
  }

  var path = require('path'),
      pkg = require('../package.json'),
      superAgent = require('superagent');

  if (superTest) {
    superTest = require(superTest);
  }

  /**
   * @constructor
   */
  function ParseSuperAgent(options) {
    options || (options = {});
    this.apiBase = 'https://api.parse.com';
    this.apiVer = options.apiVer || '1';
    this.initialize(options);
  }
  ParseSuperAgent.constructor = ParseSuperAgent;
  ParseSuperAgent.version = pkg.version;

  /**
   * Set up parse settings.
   * @param {Object.<(string|boolean)>=} options Parse application keys.
   */
  ParseSuperAgent.prototype.initialize = function(options) {
    options || (options = {});
    this.applicationId = options.applicationId || process.env.NODE_PARSE_APP_ID;
    this.restApiKey = options.restApiKey || process.env.NODE_PARSE_REST_API_KEY;
    this.masterKey = options.masterKey || process.env.NODE_PARSE_MASTER_KEY;
    if (typeof options.useMasterKey === 'boolean') {
      this.useMasterKey = options.useMasterKey;
    } else if (parseInt(process.env.NODE_USE_PARSE_MASTER_KEY || 0, 10)) {
      this.useMasterKey = true;
    } else {
      this.useMasterKey = false;
    }
    this.env = options.env || process.env.NODE_ENV;
    if (this.env === 'testing' && !!!superTest) {
      console.warn('supertest not found!!');
    }
  };

  /**
   * Parse authentication keyset.
   *
   * @param {{useMasterKey: boolean}} options A keySet flag.
   * @return {Object} Parse authentication keyset.
   */
  ParseSuperAgent.prototype.keySet = function(options) {
    options || (options = {});

    var keySet = {
      'X-Parse-Application-Id': this.applicationId,
      'X-Parse-REST-API-Key': this.restApiKey
    };
    if (options.useMasterKey === true || this.useMasterKey === true) {
      keySet['X-Parse-Master-Key'] = this.masterKey;
    }
    return keySet;
  };

  /**
   * urlPath for Parse rest api.
   * @param {...string} va_args A url path part.
   * @return {string} The url path that joined all arguments.
   */
  ParseSuperAgent.prototype.pathJoin = function(va_args) {
    return path.join.apply('', arguments);
  };

  ParseSuperAgent.prototype.request = function(method, urlPath, params, query, options) {
    var request;
    if (this.env === 'testing' && superTest) {
      request = superTest(this.apiBase)[method](urlPath);
    } else {
      request = superAgent[method](this.apiBase + urlPath);
    }
    return request
      .type('application/json')
      .set(this.keySet(options))
      .query(query || {})
      .send(params || {});
  };

  /**
   * Send the request to run Parse cloud function.
   * @param {string} name A cloud function name.
   * @param {Object=} params A request data.
   * @param {Object=} options A flags that control the object behaviour.
   * @return {Object} The request object provided by superagent.
   */
  ParseSuperAgent.prototype.runFunction = function(name, params, options) {
    var urlPath = this.pathJoin('/', this.apiVer, 'functions', name);
    return this.request('post', urlPath, params, null, options);
  };

  /**
   * Send the request to run Parse cloud background job.
   * @param {string} name A cloud background job name.
   * @param {Object=} params A request data.
   * @param {Object=} options A flags that control the object behaviour.
   * @return {Object} The request object provided by superagent.
   */
  ParseSuperAgent.prototype.runBackgroundJob = function(name, params, options) {
    var urlPath = this.pathJoin('/', this.apiVer, 'jobs', name);
    return this.request('post', urlPath, params, null, options);
  };

  /**
   * Send the batch opration request.
   * @param {Array.<Object>} params A batch opration data list.
   * @return {Object} The request object provided by superagent.
   */
  ParseSuperAgent.prototype.sendBatchRequest = function(params, options) {
    var urlPath = this.pathJoin('/', this.apiVer, 'batch');
    return this.request('post', urlPath, params, null, options);
  };

  /**
   * Send the request for Parse.com rest api.
   * @param {string} method A HTTP request method.
   * @param {string} requestPath A Parse reset api request path.
   * @param {Object=} params A request data(POST PUT).
   * @param {Object=} query A request query string(GET).
   * @param {Object=} options A flags that control the object behaviour.
   * @return {Object} The request object provided by superagent.
   */
  ParseSuperAgent.prototype.send = function(method, requestPath, params, query, options) {
    options || (options = {});
    var urlPath = requestPath;
    method = method.toLowerCase();
    method = method === 'delete' ? 'del' : method;
    return this.request(method, urlPath, params, query, options);
  };

  module.exports = ParseSuperAgent;
}(this));


examples/client.js
/**
 * client.js
 */

var parseSuperAgent = require('../lib'),
    events = require('events'),
    eventEmitter = new events.EventEmitter();

// Setup the parse keys by environment variables.
// export NODE_PARSE_APP_ID=<Application ID>
// export NODE_PARSE_REST_API_KEY=<REST API Key>
// export NODE_PARSE_MASTER_KEY=<Master Key>
// export NODE_USE_PARSE_MASTER_KEY=(0|1)

// Or the instantiate arguments.
var client = new parseSuperAgent(/*{
  applicationId: '<Application ID>',
  restApiKey: '<REST API Key>',
  masterKey: '<Master Key>',
  useMasterKey: true  # Or false
}*/);

client.runFunction('hello').end(function(error, response) {
  if (error) {
    throw error;
  }
  console.log(response.body);
  // => { result: 'Hello world!' }
});

client.sendBatchRequest({
          requests: [{
            method: 'POST',
            path: '/1/classes/GameScore',
            body: {
              score: 1337,
              playerName: "Sean Plott"
            }
          }, {
            method: 'POST',
            path: '/1/classes/GameScore',
            body: {
              score: 1338,
              playerName: "ZeroCool"
            }
          }]
}).end(function(error, response) {
  if (error) {
    throw error;
  }
  console.log(response.body);
  // => [ { success:
  // { createdAt: '2013-11-17T14:43:40.296Z',
  // objectId: 'aGFMHhdHBY' } },
  // { success:
  // { createdAt: '2013-11-17T14:43:40.308Z',
  // objectId: 'rpQbB6RXgT' } } ]
  eventEmitter.emit('batchEnd');
});

client.send('POST', '/1/classes/GameScore', {
  score:1337,
  playerName: "Sean Plott",
  cheatMode:false
}).end(function(error, response) {
  if (error) {
    throw error;
  }
  console.log(response.body);
  // => { createdAt: '2013-11-17T14:43:40.297Z',
  // objectId: 'W2tdhLOWKC' }
  eventEmitter.emit('created', response.body.objectId);
});

eventEmitter.on('created', function(objectId) {
  client
    .send('GET', '/1/classes/GameScore/' + objectId)
    .end(function(error, response) {
      if (error) {
        throw error;
      }
      console.log(response.body);
      // => { cheatMode: false,
      // playerName: 'Sean Plott',
      // score: 73453,
      // createdAt: '2013-11-17T14:43:40.297Z',
      // updatedAt: '2013-11-17T14:43:41.432Z',
      // objectId: 'W2tdhLOWKC' }
      eventEmitter.emit('fetched', response.body.objectId);
    });

  client
    .send('PUT', '/1/classes/GameScore/' + objectId, {score: 73453})
    .end(function(error, response) {
      if (error) {
        throw error;
      }
      console.log(response.body);
      // => { updatedAt: '2013-11-17T14:43:41.432Z' }
      eventEmitter.emit('updated', objectId);
    });
});

eventEmitter.on('updated', function(objectId) {
  client
    .send('DELETE', '/1/classes/GameScore/' + objectId)
    .end(function(error, response) {
      if (error) {
        throw error;
      }
      console.log(response.body);
      // => {}
      eventEmitter.emit('deleted', response.body);
    });
});

eventEmitter.on('deleted', function() {
  client
    .send('GET', '/1/classes/GameScore')
    .end(function(error, response) {
      if (error) {
        throw error;
      }
      console.log(response.body);
      // => { results:
      // [ { score: 1337,
      // playerName: 'Sean Plott',
      // createdAt: '2013-11-17T14:43:40.296Z',
      // updatedAt: '2013-11-17T14:43:40.296Z',
      // objectId: 'aGFMHhdHBY' },
      // { score: 1338,
      // playerName: 'ZeroCool',
      // createdAt: '2013-11-17T14:43:40.308Z',
      // updatedAt: '2013-11-17T14:43:40.308Z',
      // objectId: 'rpQbB6RXgT' } ] }
      //     });
      var requests = [];
      var results = response.body.results;
      for (var i = 0, l = results.length; i < l; i++) {
        requests.push({
          method: 'DELETE',
          path: '/1/classes/GameScore/' + results[i].objectId
        });
      }
      eventEmitter.emit('fetchAll', requests);
    });
});

eventEmitter.on('fetchAll', function(requests) {
  // Cleanup.
  client
    .sendBatchRequest({requests: requests})
    .end(function(error, response) {
      if (error) {
        throw error;
      }
      console.log(response.body);
      // => [ { success: true }, { success: true } ]
    });
});
